src/datanode/src/sql/copy_table_from.rs

```diff
@@ -88,58 +96,64 @@ impl SqlHandler {
                 }
             );
 
-            let stream = builder
+            let mut stream = builder
                 .build()
                 .context(error::BuildParquetRecordBatchStreamSnafu)?;
 
-            let chunk = stream
-                .try_collect::<Vec<_>>()
-                .await
-                .context(error::ReadParquetSnafu)?;
-
-            buf.extend(chunk.into_iter());
-        }
-
-        let fields = table
-            .schema()
-            .arrow_schema()
-            .fields()
-            .iter()
-            .map(|f| f.name().to_string())
-            .collect::<Vec<_>>();
-
-        // Vec<Columns>
-        let column_chunks = buf
-            .into_iter()
-            .map(|c| Helper::try_into_vectors(c.columns()).context(DataTypesSnafu))
-            .collect::<Vec<_>>();
+            // TODO(hl): make this configurable through options.
+            let pending_mem_threshold = ReadableSize::mb(32).as_bytes();
+            let mut pending_mem_size = 0;
+            let mut pending = vec![];
+
+            while let Some(r) = stream.next().await {
+                let record_batch = r.context(error::ReadParquetSnafu)?;
+                let vectors = Helper::try_into_vectors(record_batch.columns())
+                    .context(DataTypesSnafu)
+                    .context(ParseDataTypesSnafu)?;
+
+                pending_mem_size += vectors.iter().map(|v| v.memory_size()).sum::<usize>();
+
+                let columns_values = fields
+                    .iter()
+                    .cloned()
+                    .zip(vectors.into_iter())
+                    .collect::<HashMap<_, _>>();
+
+                pending.push(table.insert(InsertRequest {
+                    catalog_name: req.catalog_name.to_string(),
+                    schema_name: req.schema_name.to_string(),
+                    table_name: req.table_name.to_string(),
+                    columns_values,
+                    //TODO: support multi-regions
+                    region_number: 0,
+                }));
+
+                if pending_mem_size as u64 >= pending_mem_threshold {
```

Comment:
so it seems that once pending_mem_size is above the threshold, it will always above the threshold. then we can reduce the `sum`s in line 114. for example, introducing some boolean flag?